### ISSUE TITLE
Truncated SVD adjoint uses the s_discarded -> 0 limit for the kept/discarded coupling

### ISSUE THREAD

**Issue by yingjerkao** [tenax-lab/tenax#752]
Split out from #750 so the fix there could land with an honest scope statement. This one is a **controlled approximation, not a bug** — filing it so the limitation is recorded rather than rediscovered.

## What

`_svd_sector_backward` terms 4 and 5 handle the coupling between the kept and discarded subspaces as

```python
dM += proj_U_perp @ dU @ jnp.diag(s_inv) @ Vh_k
dM += U_k @ jnp.diag(s_inv) @ dVh @ proj_V_perp
```

`proj_U_perp = I − U_k U_kᴴ` lumps the discarded singular directions together with the true null space, and weights them by `1/s_kept`. That is the exact answer only in the limit `s_discarded → 0`; the exact coupling weight also depends on the discarded singular values themselves.

## Measured

AD vs finite differences on `truncated_svd_ad`, 6×6, kept spectrum `[4, 3, 2]` (k=3), sweeping the discarded tail. All numbers with the #750/#751 fixes in place:

| max discarded s | rel error | cos |
|---|---|---|
| 2e-1 | 2.19e-02 | 0.99976 |
| 2e-2 | 2.21e-03 | 0.999998 |
| 2e-3 | 2.21e-04 | 0.99999998 |
| 2e-4 | 2.21e-05 | 1.0000000 |
| ~1e-7 | 2.2e-08 | 1.0000000 |

The error is driven to zero by the discarded weight, roughly linearly — the signature of an approximation rather than a defect. Direction stays excellent throughout (cos > 0.9997 even at 10% discarded weight).

## Why it probably doesn't matter much in practice, and when it might

CTM truncation keeps the dominant part of the spectrum, so `s_discarded/s_kept` is normally small and the induced gradient error is correspondingly small. It becomes worth revisiting if:

- running at aggressive truncation (small χ relative to the spectrum's decay), where discarded weight is not small;
- chasing a gradient-norm convergence criterion tighter than the truncation error;
- diagnosing a residual AD-vs-FD floor — this is one of the two known contributors, the other being the complex gauge-fixing derivative in #751.

## Fix, if ever wanted

Use the full-spectrum F-matrix with zero-padded cotangents rather than the `s_r → 0` limit: the exact truncated adjoint is the *full* SVD adjoint evaluated with `dU`, `ds`, `dVh` padded with zeros on the discarded columns. Costs the full `p×p` F-matrix instead of `k×k`.

Gated (as an approximation, not an exactness claim) by `test_truncation_error_vanishes_with_discarded_weight` in `tests/test_svd_adjoint_fd_750.py`, which asserts the error keeps shrinking with the discarded weight — so a real regression in terms 4/5 would still be caught.

Related: #750, #751, #687.

> 🤖 **AI-generated issue** — written by Claude Code, posted by @yingjerkao.

**Comment by yingjerkao** [comment]
Fixed in #754 (`8751b96`) — and the premise that this had to stay an approximation was wrong.

`(U_k, s_k, Vh_k)` is a **slice** of the full SVD, so the exact VJP of the truncated map is just the full-SVD adjoint evaluated with the cotangents **zero-padded onto the discarded columns**. Padding makes the F-matrix span the whole spectrum, which is what carries the kept/discarded coupling exactly. Terms 4/5 are then left carrying only the genuine null space — which is what they were actually correct for all along.

AD vs finite differences, same sweep as the issue:

| max discarded s | before | after |
|---|---|---|
| 2e-1 | 2.19e-02 | **3.07e-09** |
| 2e-2 | 2.21e-03 | **2.51e-09** |
| 2e-3 | 2.21e-04 | **2.43e-09** |
| 2e-4 | 2.21e-05 | **2.37e-09** |

Flat at the finite-difference noise floor instead of tracking the discarded weight. `cos = 1.000000000` throughout.

**The cost concern didn't materialise.** The adjoint's inner dimension grows from `k` to `p` (`= χ·D` for the edge-split SVD, so `p/k = D`), which looked like it should cost `D²` more. It doesn't, because the forward SVD already computes the full spectrum and dominates the backward. Measured backward wall-clock at CTM-realistic shapes:

| shape | k=χ | p | approximate | exact |
|---|---|---|---|---|
| 64² | 32 | 64 | 1.78 ms | 1.87 ms |
| 128² | 32 | 128 | 8.44 ms | 11.92 ms |
| 256² | 64 | 256 | 28.80 ms | 29.38 ms |
| 256² | 128 | 256 | 34.13 ms | **28.53 ms** |
| 512² | 128 | 512 | 121.00 ms | 123.34 ms |

Within noise, and faster at one point. So there was no accuracy/cost trade to make.

One subtlety worth recording: the rank mask now covers only **kept**-but-zero columns. Discarded indices are deliberately left unmasked — their padded cotangents are identically zero, so they inject nothing gauge-arbitrary, while their `1/(s_j² − s_r²)` entries are exactly the coupling this change adds. Masking them would have silently restored the approximation this fixes.

New gates in `tests/test_svd_adjoint_fd_750.py`: exactness at every discarded weight (parametrized), non-square input in both orientations (where the null-space terms are non-trivial), and rank-deficient input. `test_ad_utils`'s independent oracle is extended to the padded form.

Verified: `-m core` 1298 passed / 8 skipped; `test_ad_utils` + rank-aware 67 passed / 1 skipped / 3 xfailed; `test_svd_adjoint_fd_750` 21 passed.

> 🤖 **AI-generated comment** — written by Claude Code, posted by @yingjerkao.
